File: backend/app/integrations/moodle_client.py
"""Moodle Web Services API client."""
import logging
from typing import Any, Optional
from urllib.parse import urljoin

import httpx

logger = logging.getLogger(__name__)


class MoodleAPIError(Exception):
    """Moodle API error."""

    pass


class MoodleClient:
    """Client for Moodle Web Services REST API."""

    def __init__(self, moodle_url: str, api_token: str):
        """Initialize Moodle client.

        Args:
            moodle_url: Base URL of Moodle instance (e.g., 'https://moodle.example.com')
            api_token: Moodle web services token
        """
        self.moodle_url = moodle_url.rstrip("/")
        self.api_token = api_token
        self.webservice_url = urljoin(
            self.moodle_url, "/webservice/rest/server.php"
        )
        self.timeout = 30.0

    async def _call_api(
        self,
        function: str,
        params: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Call Moodle Web Services API.

        Args:
            function: Moodle web service function name
            params: Additional parameters for the API call

        Returns:
            API response as dictionary

        Raises:
            MoodleAPIError: If API call fails
        """
        request_params = {
            "wstoken": self.api_token,
            "wsfunction": function,
            "moodlewsrestformat": "json",
        }

        if params:
            request_params.update(params)

        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                response = await client.get(self.webservice_url, params=request_params)
                response.raise_for_status()
                data = response.json()

                # Check for Moodle API errors
                if isinstance(data, dict) and "exception" in data:
                    error_msg = data.get("message", "Unknown Moodle API error")
                    logger.error(f"Moodle API error: {error_msg}")
                    raise MoodleAPIError(error_msg)

                return data

        except httpx.HTTPError as e:
            logger.error(f"HTTP error calling Moodle API: {e}")
            raise MoodleAPIError(f"HTTP error: {str(e)}")
        except Exception as e:
            logger.error(f"Unexpected error calling Moodle API: {e}")
            raise MoodleAPIError(f"Unexpected error: {str(e)}")

    async def test_connection(self) -> bool:
        """Test connection to Moodle API.

        Returns:
            True if connection is successful
        """
        try:
            result = await self._call_api("core_webservice_get_site_info")
            logger.info(f"Successfully connected to Moodle: {result.get('sitename')}")
            return True
        except MoodleAPIError:
            return False

    async def get_courses(self) -> list[dict[str, Any]]:
        """Get list of courses.

        Returns:
            List of course dictionaries
        """
        result = await self._call_api("core_course_get_courses")
        return result if isinstance(result, list) else []

    async def get_course_by_id(self, course_id: int) -> dict[str, Any]:
        """Get course details by ID.

        Args:
            course_id: Course ID

        Returns:
            Course details dictionary
        """
        result = await self._call_api(
            "core_course_get_courses_by_field",
            {"field": "id", "value": course_id},
        )
        courses = result.get("courses", [])
        if courses:
            return courses[0]
        raise MoodleAPIError(f"Course {course_id} not found")

    async def get_quizzes_by_course(self, course_id: int) -> list[dict[str, Any]]:
        """Get quizzes for a course.

        Args:
            course_id: Course ID

        Returns:
            List of quiz dictionaries
        """
        result = await self._call_api(
            "mod_quiz_get_quizzes_by_courses",
            {"courseids[0]": course_id},
        )
        return result.get("quizzes", [])

    async def get_quiz_by_id(self, quiz_id: int) -> dict[str, Any]:
        """Get quiz details by ID.

        Args:
            quiz_id: Quiz ID

        Returns:
            Quiz details dictionary
        """
        result = await self._call_api(
            "mod_quiz_get_quiz_by_courseid",
            {"quizid": quiz_id},
        )
        return result

    async def get_quiz_questions(self, quiz_id: int) -> list[dict[str, Any]]:
        """Get questions for a quiz.

        Args:
            quiz_id: Quiz ID

        Returns:
            List of question dictionaries
        """
        # Note: This requires mod_quiz_get_quiz_access_information and
        # mod_quiz_get_attempt_data functions to be enabled
        try:
            result = await self._call_api(
                "mod_quiz_get_quiz_by_courseid",
                {"quizid": quiz_id},
            )
            # Extract questions from quiz structure
            # This is a simplified version - actual implementation depends on Moodle version
            return result.get("questions", [])
        except MoodleAPIError:
            logger.warning(f"Could not fetch questions for quiz {quiz_id}")
            return []

    async def get_user_by_id(self, user_id: int) -> dict[str, Any]:
        """Get user details by ID.

        Args:
            user_id: User ID

        Returns:
            User details dictionary
        """
        result = await self._call_api(
            "core_user_get_users_by_field",
            {"field": "id", "values[0]": user_id},
        )
        users = result if isinstance(result, list) else []
        if users:
            return users[0]
        raise MoodleAPIError(f"User {user_id} not found")

    async def get_enrolled_users(self, course_id: int) -> list[dict[str, Any]]:
        """Get enrolled users in a course.

        Args:
            course_id: Course ID

        Returns:
            List of user dictionaries
        """
        result = await self._call_api(
            "core_enrol_get_enrolled_users",
            {"courseid": course_id},
        )
        return result if isinstance(result, list) else []

    async def update_grades(
        self,
        course_id: int,
        user_id: int,
        grade_item: str,
        grade: float,
    ) -> bool:
        """Update user's grade in Moodle.

        Args:
            course_id: Course ID
            user_id: User ID
            grade_item: Grade item name
            grade: Grade value

        Returns:
            True if successful
        """
        try:
            await self._call_api(
                "core_grades_update_grades",
                {
                    "source": "lms_practice_integration",
                    "courseid": course_id,
                    "component": "mod_quiz",
                    "activityid": 0,  # Will be set based on actual quiz
                    "itemnumber": 0,
                    "grades[0][studentid]": user_id,
                    "grades[0][grade]": grade,
                },
            )
            logger.info(
                f"Updated grade for user {user_id} in course {course_id}: {grade}"
            )
            return True
        except MoodleAPIError as e:
            logger.error(f"Failed to update grades: {e}")
            return False

    async def get_question_bank(self, course_id: int) -> list[dict[str, Any]]:
        """Get questions from course question bank.

        Args:
            course_id: Course ID

        Returns:
            List of question dictionaries

        Note:
            This requires additional Moodle plugins or custom web services.
            Implementation depends on available Moodle web services.
        """
        # This is a placeholder - actual implementation depends on Moodle setup
        logger.warning(
            "get_question_bank requires custom Moodle web service or plugin"
        )
        return []

    async def create_quiz_attempt(
        self,
        quiz_id: int,
        user_id: int,
    ) -> dict[str, Any]:
        """Create a new quiz attempt for a user.

        Args:
            quiz_id: Quiz ID
            user_id: User ID

        Returns:
            Attempt details dictionary
        """
        result = await self._call_api(
            "mod_quiz_start_attempt",
            {
                "quizid": quiz_id,
                "userid": user_id,
            },
        )
        return result

    async def submit_quiz_answer(
        self,
        attempt_id: int,
        question_id: int,
        answer: str,
    ) -> bool:
        """Submit answer for a quiz question.

        Args:
            attempt_id: Attempt ID
            question_id: Question ID
            answer: Student's answer

        Returns:
            True if successful
        """
        try:
            await self._call_api(
                "mod_quiz_process_attempt",
                {
                    "attemptid": attempt_id,
                    "data[0][name]": f"q{question_id}:answer",
                    "data[0][value]": answer,
                },
            )
            return True
        except MoodleAPIError as e:
            logger.error(f"Failed to submit answer: {e}")
            return False

File: backend/utils/MoodleClient.php
<?php
/**
 * Moodle API Client
 * Handles communication with Moodle 3.7 Web Services
 */

class MoodleClient {
    private $moodleUrl;
    private $token;
    private $service;

    public function __construct($config) {
        $this->moodleUrl = rtrim($config['moodle']['url'], '/');
        $this->token = $config['moodle']['token'];
        $this->service = $config['moodle']['service'];
    }

    /**
     * Call Moodle Web Service API
     *
     * @param string $function The Moodle function to call
     * @param array $params Function parameters
     * @return mixed API response
     */
    private function callApi($function, $params = []) {
        $url = $this->moodleUrl . '/webservice/rest/server.php';

        $postData = [
            'wstoken' => $this->token,
            'wsfunction' => $function,
            'moodlewsrestformat' => 'json',
        ];

        // Merge function parameters
        $postData = array_merge($postData, $params);

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($postData));
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, 30);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);

        if (curl_errno($ch)) {
            $error = curl_error($ch);
            curl_close($ch);
            throw new Exception('Moodle API request failed: ' . $error);
        }

        curl_close($ch);

        if ($httpCode !== 200) {
            throw new Exception('Moodle API returned HTTP ' . $httpCode);
        }

        $data = json_decode($response, true);

        if (isset($data['exception'])) {
            throw new Exception('Moodle API error: ' . $data['message']);
        }

        return $data;
    }

    /**
     * Get quiz information
     *
     * @param int $quizId Quiz ID
     * @return array Quiz data
     */
    public function getQuiz($quizId) {
        return $this->callApi('mod_quiz_get_quizzes_by_courses', [
            'courseids[0]' => 0, // Will get all courses, then filter
        ]);
    }

    /**
     * Get quiz attempt for a user
     *
     * @param int $quizId Quiz ID
     * @param int $userId Moodle user ID
     * @return array Attempt data
     */
    public function getUserAttempt($quizId, $userId) {
        return $this->callApi('mod_quiz_get_user_attempts', [
            'quizid' => $quizId,
            'userid' => $userId,
        ]);
    }

    /**
     * Get quiz questions
     *
     * @param int $attemptId Attempt ID
     * @return array Questions data
     */
    public function getQuizQuestions($attemptId) {
        return $this->callApi('mod_quiz_get_attempt_data', [
            'attemptid' => $attemptId,
        ]);
    }

    /**
     * Submit quiz answer
     *
     * @param int $attemptId Attempt ID
     * @param int $questionId Question ID
     * @param string $answer Answer text
     * @return array Submit result
     */
    public function submitAnswer($attemptId, $questionId, $answer) {
        return $this->callApi('mod_quiz_process_attempt', [
            'attemptid' => $attemptId,
            'data[' . $questionId . '_answer]' => $answer,
        ]);
    }

    /**
     * Finish quiz attempt
     *
     * @param int $attemptId Attempt ID
     * @return array Finish result
     */
    public function finishAttempt($attemptId) {
        return $this->callApi('mod_quiz_process_attempt', [
            'attemptid' => $attemptId,
            'finishattempt' => 1,
        ]);
    }

    /**
     * Get user info
     *
     * @param int $userId Moodle user ID
     * @return array User data
     */
    public function getUser($userId) {
        $result = $this->callApi('core_user_get_users_by_field', [
            'field' => 'id',
            'values[0]' => $userId,
        ]);

        return isset($result[0]) ? $result[0] : null;
    }

    /**
     * Get course info
     *
     * @param int $courseId Course ID
     * @return array Course data
     */
    public function getCourse($courseId) {
        $result = $this->callApi('core_course_get_courses', [
            'options[ids][0]' => $courseId,
        ]);

        return isset($result[0]) ? $result[0] : null;
    }

    /**
     * Send grade to Moodle gradebook
     *
     * @param int $userId User ID
     * @param int $itemId Grade item ID
     * @param float $grade Grade value
     * @return array Result
     */
    public function updateGrade($userId, $itemId, $grade) {
        return $this->callApi('core_grades_update_grades', [
            'source' => 'power_candle',
            'courseid' => 0, // Will be determined by item
            'component' => 'mod_quiz',
            'activityid' => $itemId,
            'itemnumber' => 0,
            'grades[0][studentid]' => $userId,
            'grades[0][grade]' => $grade,
        ]);
    }

    /**
     * Parse Moodle question to Power Candle format
     *
     * @param array $moodleQuestion Moodle question data
     * @return array|null Power Candle problem data or null if not parseable
     */
    public function parseProblemFromMoodle($moodleQuestion) {
        // Expected Moodle question format:
        // Question text: "Calculate: log_2 8 = ?"
        // or custom question field for Power Candle

        $questionText = $moodleQuestion['questiontext'] ?? '';

        // Try to parse logarithm from question text
        // Pattern: log_(\d+)\s+(\d+)
        if (preg_match('/log[_₍]?(\d+)[₎]?\s+(\d+)/', $questionText, $matches)) {
            $base = (int)$matches[1];
            $result = (int)$matches[2];

            // Calculate correct answer
            $calculator = new LogarithmCalculator();
            try {
                $correctAnswer = $calculator->calculate($base, $result);

                return [
                    'moodle_question_id' => $moodleQuestion['id'] ?? 0,
                    'moodle_quiz_id' => $moodleQuestion['quizid'] ?? null,
                    'problem_type' => 'calculate',
                    'base' => $base,
                    'result' => $result,
                    'correct_answer' => $correctAnswer,
                    'question_text' => $questionText,
                    'difficulty' => $this->estimateDifficulty($base, $correctAnswer),
                ];
            } catch (Exception $e) {
                error_log('Failed to parse Moodle question: ' . $e->getMessage());
                return null;
            }
        }

        return null;
    }

    /**
     * Estimate difficulty based on base and power
     *
     * @param int $base The base
     * @param int $power The power
     * @return int Difficulty level (1-5)
     */
    private function estimateDifficulty($base, $power) {
        // Difficulty increases with higher powers and larger bases
        $score = $power + ($base > 3 ? 1 : 0);

        if ($score <= 2) return 1; // Easy
        if ($score <= 3) return 2; // Medium-Easy
        if ($score <= 4) return 3; // Medium
        if ($score <= 5) return 4; // Medium-Hard
        return 5; // Hard
    }

    /**
     * Sync problems from Moodle quiz
     *
     * @param int $quizId Moodle quiz ID
     * @param Database $db Database instance
     * @return array Sync result
     */
    public function syncProblemsFromQuiz($quizId, $db) {
        try {
            // Get quiz data
            $quiz = $this->getQuiz($quizId);

            // Note: This is a simplified version
            // In production, you would need to get a specific attempt or create one
            // to access the questions

            $synced = 0;
            $skipped = 0;
            $errors = [];

            // For each question in the quiz
            // (Implementation depends on Moodle's quiz structure)

            return [
                'success' => true,
                'synced' => $synced,
                'skipped' => $skipped,
                'errors' => $errors,
            ];
        } catch (Exception $e) {
            return [
                'success' => false,
                'error' => $e->getMessage(),
            ];
        }
    }

    /**
     * Test connection to Moodle
     *
     * @return bool True if connection successful
     */
    public function testConnection() {
        try {
            $result = $this->callApi('core_webservice_get_site_info');
            return isset($result['sitename']);
        } catch (Exception $e) {
            error_log('Moodle connection test failed: ' . $e->getMessage());
            return false;
        }
    }

    /**
     * Get Moodle site info
     *
     * @return array Site information
     */
    public function getSiteInfo() {
        return $this->callApi('core_webservice_get_site_info');
    }
}
